### PR TITLE
Adds support for Windows worker nodes

### PR DESCRIPTION
**Pull request by gab-satchi**
- Adds windows flavour to template generator
- Adds OS field to vsphere machines

**What this PR does / why we need it**:

**Special notes for your reviewer**:
Relies on an image built from this image builder [PR](https://github.com/kubernetes-sigs/image-builder/pull/402)

**Release note**:
```release-note
Support Windows worker nodes
```

### CODE REVIEW

File: controllers/vspheremachine_controller.go [hunk 1]
```diff
@@ -391,10 +391,21 @@ func (r machineReconciler) reconcileNormal(ctx *context.MachineContext) (reconci
 
 func (r machineReconciler) reconcileNormalPre7(ctx *context.MachineContext, vsphereVM *infrav1.VSphereVM) (runtime.Object, error) {
 	// Create or update the VSphereVM resource.
+	vsphereVMName := ctx.VSphereMachine.Name
```
Comment (yastij) [inline controllers/vspheremachine_controller.go:394]: shouldn't this be `ctx.Machine.Name` ?

File: controllers/vspheremachine_controller.go [hunk 2]
```diff
@@ -391,10 +391,21 @@ func (r machineReconciler) reconcileNormal(ctx *context.MachineContext) (reconci
...
+			ctx.Logger.Error(err, "unable to generate hostname for vSphereVM", "vm", ctx.VSphereMachine.Name)
+			return nil, err
+		}
+	}
+
 	vm := &infrav1.VSphereVM{
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace: ctx.VSphereMachine.Namespace,
-			Name:      ctx.Machine.Name,
+			Name:      vsphereVMName,
```
Comment (yastij) [inline controllers/vspheremachine_controller.go:408]: another approach is to push this to the mutating webhook ? it would then decide if it needs a mutation or not (I'm mainly thinking about the case where we'd want to delegate VM creation to another component in the future)

**Review by yastij — COMMENTED** [review]
I know it's a wip, so just high level review on the approach

File: controllers/vspheremachine_controller.go (hunk 1 above)
Comment (gab-satchi) [reply controllers/vspheremachine_controller.go:394]: yes. my mistake

File: controllers/vspheremachine_controller.go (hunk 2 above)
Comment (gab-satchi) [reply controllers/vspheremachine_controller.go:408]: webhook makes more sense. will update

File: packaging/flavorgen/flavors/generators.go [hunk 3]
```diff
@@ -505,6 +579,45 @@ func newMachineDeployment(cluster clusterv1.Cluster, machineTemplate infrav1.VSp
...
+			Kind:       typeToKind(&clusterv1.MachineDeployment{}),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s%s-windows", clusterNameVar, machineDeploymentNameSuffix),
+			Labels:    clusterLabels(),
+			Namespace: namespaceVar,
+		},
+		Spec: clusterv1.MachineDeploymentSpec{
+			ClusterName: clusterNameVar,
+			Replicas:    pointer.Int32Ptr(int32(555)),
```
Comment (jayunit100) [inline packaging/flavorgen/flavors/generators.go:595]: 555? whatys that all bout ? :)

File: api/v1alpha3/vspherevm_webhook.go [hunk 4]
```diff
@@ -35,6 +36,19 @@ func (r *VSphereVM) SetupWebhookWithManager(mgr ctrl.Manager) error {
 }
 
 // +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha3-vspherevm,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=vspherevms,versions=v1alpha3,name=validation.vspherevm.infrastructure.x-k8s.io,sideEffects=None
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-infrastructure-cluster-x-k8s-io-v1alpha3-vspherevm,mutating=true,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=vspherevms,versions=v1alpha3,name=default.vspherevm.infrastructure.x-k8s.io,sideEffects=None
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *VSphereVM) Default() {
+	// Windows hostnames must be < 16 characters in length
+	if r.Spec.OS == Windows && len(r.Name) > 15 {
+		name, err := base36TruncatedHash(r.Name, 15)
```
Comment (jayunit100) [inline api/v1alpha3/vspherevm_webhook.go:45]: just thinking, for UX, could we have the last 5 chars be random, and the first part == the windows VM name so theyre easy to read?

File: packaging/flavorgen/flavors/generators.go (hunk 3 above)
Comment (gab-satchi) [reply packaging/flavorgen/flavors/generators.go:595]: just a placeholder that would then get replaced with the env var `WINDOWS_WORKER_MACHINE_COUNT`

File: packaging/flavorgen/flavors/flavors.go [hunk 5]
```diff
@@ -88,6 +88,39 @@ func MultiNodeTemplateWithKubeVIP() []runtime.Object {
 	return append(MultiNodeTemplate, crsResources...)
 }
 
+// MultiNodeTemplateWithWindows generates templates for a hybrid cluster with Linux and Windows worker nodes
+func MultiNodeTemplateWithWindows() []runtime.Object {
+	vsphereCluster := newVSphereCluster(nil)
+	linuxMachineTemplate := newVSphereLinuxMachineTemplate()
+	controlPlane := newKubeadmControlplane(445, linuxMachineTemplate, newKubeVIPFiles())
```
Comment (jayunit100) [inline packaging/flavorgen/flavors/flavors.go:95]: not important but curious what this # is maybe a comment for it?

File: packaging/flavorgen/flavors/flavors.go [hunk 6]
```diff
@@ -88,6 +88,39 @@ func MultiNodeTemplateWithKubeVIP() []runtime.Object {
...
+	cluster := newCluster(vsphereCluster, &controlPlane)
+	linuxMachineDeployment := newLinuxMachineDeployment(cluster, linuxMachineTemplate, linuxKubeadmJoinTemplate)
+	windowsMachineTemplate := newVSphereWindowsMachineTemplate()
+	windowsKubeadmJoinTemplate := newWindowsKubeadmConfigTemplate()
+	windowsMachineDeployment := newWindowsMachineDeployment(cluster, windowsMachineTemplate, windowsKubeadmJoinTemplate)
+
+	clusterResourceSet := newClusterResourceSet(cluster)
+	crsResources := createCrsResourceObjects(&clusterResourceSet, vsphereCluster, cluster)
+
+	// removing Storage config so the cluster controller is not going not install CSI (it is installed by the clusterResourceSet)
```
Comment (jayunit100) [inline packaging/flavorgen/flavors/flavors.go:106]: interesting, so, is that asymmetrical WRT the linux template? or is this all changing.  

i know vsphere stroage driver doesnt **yet** support windows, is that the reasoning for setting this to nil ? seems like in an ideal world you wouldnt have to do a one-off thing for the windows case.

File: api/v1alpha3/vspherevm_webhook.go (hunk 4 above)
Comment (gab-satchi) [reply api/v1alpha3/vspherevm_webhook.go:45]: yeah that could work. we currently use the machine deployment name `<cluster-name>-windows` where the default linux machine deployment is just `<cluster-name>`. With just 15 characters, it's hard to fit something meaningful that would reflect the cluster it belongs to and differentiate Windows while staying unique

File: packaging/flavorgen/flavors/flavors.go (hunk 5 above)
Comment (gab-satchi) [reply packaging/flavorgen/flavors/flavors.go:95]: ah that's another one of those placeholder replicas :) I'm not sure why the high numbers were used, just grabbed it from the Linux templates

File: packaging/flavorgen/flavors/generators.go (hunk 3 above)
Comment (randomvariable) [reply packaging/flavorgen/flavors/generators.go:595]: Yup. It's just to make it a valid value before being replaced with a string templating variable in the final output.

To be fair, we wrote this whole flavour generator thing a while ago when there was a lot of boilerplating needed for vsphere clusters, particularly around CSI and CPI. Could potentially get rid of it in a future PR.

File: api/v1alpha3/vspherevm_webhook.go (hunk 4 above)
Comment (randomvariable) [reply api/v1alpha3/vspherevm_webhook.go:45]: Did the NETBIOS limitation turn out to be important?

File: packaging/flavorgen/flavors/flavors.go (hunk 6 above)
Comment (gab-satchi) [reply packaging/flavorgen/flavors/flavors.go:106]: no with these changes, it still attempts the csi setup but through cluster resource sets. Depending on when vsphere storage driver support is expected for Windows, I can remove it from here before merge

File: packaging/flavorgen/flavors/generators.go [hunk 7]
```diff
@@ -312,6 +379,13 @@ func defaultNodeRegistrationOptions() kubeadmv1beta1.NodeRegistrationOptions {
 	}
 }
 
+func windowsNodeRegistrationOptions() kubeadmv1beta1.NodeRegistrationOptions {
+	return kubeadmv1beta1.NodeRegistrationOptions{
+		Name:             "{{ ds.meta_data.hostname }}",
```
Comment (jayunit100) [inline packaging/flavorgen/flavors/generators.go:384]: is this a capi thing ? what substitutes thos `{{` ? maybe a comment on this guy just to clarifiy

File: config/webhook/manifests.yaml [hunk 8]
```diff
@@ -1,4 +1,32 @@
 
+---
```
Comment (jayunit100) [inline config/webhook/manifests.yaml:2]: whats specific to windows here? again just thinking out loud as i go through it

File: packaging/flavorgen/flavors/flavors.go (hunk 5 above)
Comment (jayunit100) [reply packaging/flavorgen/flavors/flavors.go:95]: imo 123 would be less likely to cause a red -herring :) 445 almost seems to imply a permissions setting or somethin

File: packaging/flavorgen/flavors/flavors.go (hunk 6 above)
Comment (jayunit100) [reply packaging/flavorgen/flavors/flavors.go:106]: oh duh, we ALWAYS set this to nil, even in the MultipleNodeTempalteWKubeVip situation, so its just boilerplate i gueess

File: config/crd/bases/infrastructure.cluster.x-k8s.io_haproxyloadbalancers.yaml [hunk 9]
```diff
@@ -231,6 +231,10 @@ spec:
                       value in the template from which the virtual machine is cloned.
                     format: int32
                     type: integer
+                  os:
```
Comment (jayunit100) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_haproxyloadbalancers.yaml:234]: why is this needed for haproxy ? wouldnt it always be linux ? or just a boilerplate/generatino thing?

File: config/crd/bases/infrastructure.cluster.x-k8s.io_vspheremachines.yaml [hunk 10]
```diff
@@ -491,6 +491,10 @@ spec:
                   value in the template from which the virtual machine is cloned.
                 format: int32
                 type: integer
+              os:
```
Comment (jayunit100) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_vspheremachines.yaml:494]: i seem to have a cluster operating with this but i havent added `os` to every CRD (vspheremachinetemplates, vspheremachines, vspherevms).

is OS actually needed in all 3 of these YAMLs ?  or just "vspheremachinetemplates" ? `os:`

File: packaging/flavorgen/flavors/generators.go (hunk 7 above)
Comment (randomvariable) [reply packaging/flavorgen/flavors/generators.go:384]: it's jinja templating in cloud-init.

File: packaging/flavorgen/flavors/flavors.go (hunk 5 above)
Comment (randomvariable) [reply packaging/flavorgen/flavors/flavors.go:95]: it's file mode, so perms.

File: config/webhook/manifests.yaml (hunk 8 above)
Comment (randomvariable) [reply config/webhook/manifests.yaml]: just the auto-generation at work

File: config/crd/bases/infrastructure.cluster.x-k8s.io_haproxyloadbalancers.yaml (hunk 9 above)
Comment (randomvariable) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_haproxyloadbalancers.yaml:234]: just the auto-generation at work

File: config/crd/bases/infrastructure.cluster.x-k8s.io_vspheremachines.yaml (hunk 10 above)
Comment (randomvariable) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_vspheremachines.yaml:494]: just the auto-generation at work. CAPV probably needs similar clean up to https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1506 , but can ignore for now.

**Comment by randomvariable** [comment]
@gab-satchi , if we ignore NetBIOS support, with image-builder changes, is this PR still needed?

**Comment by gab-satchi** [comment]
won't be required. The controller logic won't need changing at all but the Windows flavour template generation can still be used

**Comment by randomvariable** [comment]
@gab-satchi cool. maybe worth splitting the flavour stuff out into a separate PR, and then we can decide based on the outcome of https://github.com/kubernetes-sigs/cluster-api/issues/2217

**Comment by rhockenbury** [comment]
If this work is still active, it might be worthwhile to solidify the naming pattern for windows machines across cluster-api providers.  Perhaps consider something similiar to capz - https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1313.
